composer/trainer/checkpoint.py

```diff
@@ -415,23 +443,39 @@ def save_checkpoint(self, state: State, seed: int, device: Device) -> None:
             if is_module_deepspeed(state.model):
                 model = cast("deepspeed.DeepSpeedEngine", state.model)
                 model.save_checkpoint(tmpdir, _DEEPSPEED_TAG)
+                # ensure that deepspeed checkpoints are saved in an archive
+                self.file_extension, self.write_mode = _ensure_archive(file_extension=self.file_extension,
+                                                                       write_mode=self.write_mode)
 
+            composer_states_filepath = os.path.join(tmpdir, _COMPOSER_STATES_FILENAME)
             if dist.get_global_rank() == 0:
                 # only rank 0 saves checkpoints
 
                 # we add the state only on rank 0 since other processes don't have loggers to serialize
                 # it should be the same across all ranks. per-rank state not stored
                 state_dict['state'] = state.state_dict()
 
-                composer_states_filepath = os.path.join(tmpdir, _COMPOSER_STATES_FILENAME)
                 with open(composer_states_filepath, 'xb') as f:
                     torch.save(state_dict, f)
 
-            checkpoint_archive_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
-            with tarfile.open(checkpoint_archive_filepath, self.write_mode) as tarball:
-                tarball.add(tmpdir, arcname="")  # add files flat to the tarball
-
-            log.info(f'Trainer checkpoint saved to {checkpoint_archive_filepath}')
+            checkpoint_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
+            if _is_pt_file(checkpoint_filepath):
```

Comment:
I wanted to avoid nested if statements but intentionally using them here for clarity that there is a unique codepath for `_is_pt_file` being true

Reply:
Should the rank zero check be conditioned on whether one is using deepspeed, not whether the checkpoint is ending in  `.pt`? Thinking that if one is using compression and not deepspeed, the only rank zero should make a checkpoint tarball.

Reply:
If someone is using compression you still want to tarball the files though because the original file is written as a .pt file regardless. It is true that with compression + without deepspeed all ranks will be trying to make the tarball but that doesn't do anything bad in terms of performance and makes the code cleaner. 

I thought about making the `else` an `elif is_module_deepspeed(state.model) or dist.get_global_rank() == 0` but that just seemed more confusing from a readability perspective.

I'm not super strongly opinionated about that though so lmk if you want it changed

Reply:
Agree it's not a performance hit, but I think it would be confusing if someone enabled compression and now started to see a bunch of tarballs showing up on non-zero ranks. They may not realize that these can safely be ignored. I know this PR didn't introduce this behavior, but it would be great if we could clean it up here.